### PR TITLE
feat(extract): validate connection extraction against ecommerce-demo-app ground truth

### LINKED ISSUE #241: [M3-D3.1/D3.2/D3.3] Validate connection extraction against ecommerce-demo-app ground truth
### Context
**PRD:** \`docs/project/PRD/active/PRD-phase-12-connection-detection.md\` — M3 D3.1, D3.2, D3.3

End-to-end validation that connection extraction produces correct results against the ecommerce-demo-app. Requires defining expected connections independently of extraction output, refactoring the demo app to use EventPublisherDef pattern, and comparing extraction output to ground truth.

### Traceability
- **PRD Section:** M3-D3.1 (Ground truth), M3-D3.2 (Refactor event publishing), M3-D3.3 (Full extraction validation)
- **Functional Requirements:** 100% accuracy against ground truth
- **Success Criteria:** SC-6 (100% connection extraction, zero false positives, zero false negatives)

### Acceptance Criteria
- ✓ Ground truth file created listing all expected component-to-component connections in demo app
- ✓ Ground truth defined independently (not derived from extraction output)
- ✓ Ground truth format: JSON or YAML matching Riviere schema links structure
- ✓ Ground truth validates against Riviere schema Link array type
- ✓ Demo app generic publishEvent() replaced with typed publish methods following EventPublisherDef pattern
- ✓ Demo app compiles with zero TypeScript errors after refactor
- ✓ Demo app ESLint rules pass (including new publish method rule from D2.1)
- ✓ Full extraction run against demo app
- ✓ Comparison on (source, target, type) fields — zero false positives, zero false negatives
- ✓ Performance baselines recorded in \`docs/architecture/performance/phase-12-baselines.md\`

### Edge Case Scenarios
- Demo app has event published but no subscriber → link from publisher to event exists, no subscriber link (not an error)
- Demo app has event handler subscribing to event from external system → no link (external events not in components)
- Ground truth includes transitive connections through non-components → extraction must produce matching links

### Implementation Guidelines
- **Architecture:** Changes to ecommerce-demo-app (external repo) and baseline documentation per §9.2.3, §9.2.4
- **Firm constraints:** Ground truth must exist BEFORE running extraction (§D3.1)
- **Flexible decisions:** Ground truth file format (JSON or YAML), location within demo app
- Manually trace all component-to-component connections in demo app to build ground truth
- Refactor demo app to implement EventPublisherDef for each event publisher
- Comparison script/test that diffs extraction output against ground truth on (source, target, type)
- Record timing in \`docs/architecture/performance/phase-12-baselines.md\`

### Embedded Reasoning
**Why:** Without ground truth validation, we cannot know if extraction is correct. This is the definitive proof that connection detection works end-to-end.
**What:** Ground truth definition, demo app refactor to Golden Path patterns, automated comparison test.
**How:** Manually audit demo app connections → write ground truth file → refactor publishEvent() → run extraction → diff against ground truth.

### Testing Strategy
- **Unit:** Ground truth file validates against schema.
- **Integration:** Full extraction pipeline run against demo app, output compared to ground truth. Zero diff = pass.
- **Edge Cases:** Transitive connections, events with no subscribers, multiple publishers for same event.

### Dependencies
Depends on #237 #238 #239 #240

### Verification
\`\`\`bash
pnpm nx test riviere-cli -- --testNamePattern "demo-app-validation"
# Or dedicated validation script
\`\`\`

---

## Amendment: Call Graph Engine Enhancements (3 Root Causes Fixed)

During ecommerce-demo-app validation, 57/77 connections were detected (74%). Analysis identified 20 missing connections across 6 gap categories sharing 3 root causes in the call graph engine. All 3 root causes have been fixed.

### Root Causes & Fixes

**RC1: Method-level source components skipped** (unlocked 7 connections: Gaps 1+2)
- `findClassInProject` matched `class.getStartLineNumber() === component.location.line`. Method-level components (e.g., `orders:api:handle`) point to the method line, not the class. Added `findMethodLevelComponent` to trace-calls.ts that searches for methods matching the component location.

**RC2: No container-aware component lookup** (unlocked 12 connections: Gaps 3+4+6)
- `ComponentIndex.getComponentByTypeName('Order')` failed because components are indexed by method name (`begin`, `cancel`), not class name (`Order`). Added `resolveContainerMethod` to call-graph-shared.ts that finds the class, finds the method, and uses `getComponentByLocation` to check if it's a component.

**RC3: Standalone function components not traced** (unlocked 1 connection: Gap 5)
- `buildCallGraph` only processed class methods. Added `findFunctionInProject` and `processFunction` for standalone function components like background jobs.

### Files Changed
- `call-graph/trace-calls.ts` — Refactored with TraceContext pattern; added `findMethodLevelComponent`, `findFunctionInProject`, container lookup in trace path
- `call-graph/build-call-graph.ts` — Added method-level, container, and function support in main loop
- `call-graph/call-graph-shared.ts` — Added `resolveContainerMethod`
- `call-graph/build-call-graph.gaps.spec.ts` — New test file: 10 tests covering all 3 root causes
- `detect-connections.ts` — Fixed pre-existing bug: strict mode was hardcoded to false

### Review Focus
1. **TraceContext refactor** in trace-calls.ts — extracted from monolithic `traceCallsInBody` to reduce cyclomatic complexity
2. **Container lookup correctness** — `resolveContainerMethod` iterates project source files; first matching class name wins
3. **Self-link prevention** — container lookup reuses `componentIdentity` check in both processCallExpression and traceCallExpression
4. **Test isolation** — shared project means class names must be unique across tests (addressed in gap spec)

### PR DESCRIPTION
**Pull request by NTCoding**
Closes #241

### Context
**PRD:** \`docs/project/PRD/active/PRD-phase-12-connection-detection.md\` — M3 D3.1, D3.2, D3.3

End-to-end validation that connection extraction produces correct results against the ecommerce-demo-app. Requires defining expected connections independently of extraction output, refactoring the demo app to use EventPublisherDef pattern, and comparing extraction output to ground truth.

### Traceability
- **PRD Section:** M3-D3.1 (Ground truth), M3-D3.2 (Refactor event publishing), M3-D3.3 (Full extraction validation)
- **Functional Requirements:** 100% accuracy against ground truth
- **Success Criteria:** SC-6 (100% connection extraction, zero false positives, zero false negatives)

### Acceptance Criteria
- ✓ Ground truth file created listing all expected component-to-component connections in demo app
- ✓ Ground truth defined independently (not derived from extraction output)
- ✓ Ground truth format: JSON or YAML matching Riviere schema links structure
- ✓ Ground truth validates against Riviere schema Link array type
- ✓ Demo app generic publishEvent() replaced with typed publish methods following EventPublisherDef pattern
- ✓ Demo app compiles with zero TypeScript errors after refactor
- ✓ Demo app ESLint rules pass (including new publish method rule from D2.1)
- ✓ Full extraction run against demo app
- ✓ Comparison on (source, target, type) fields — zero false positives, zero false negatives
- ✓ Performance baselines recorded in \`docs/architecture/performance/phase-12-baselines.md\`

### Edge Case Scenarios
- Demo app has event published but no subscriber → link from publisher to event exists, no subscriber link (not an error)
- Demo app has event handler subscribing to event from external system → no link (external events not in components)
- Ground truth includes transitive connections through non-components → extraction must produce matching links

### Implementation Guidelines
- **Architecture:** Changes to ecommerce-demo-app (external repo) and baseline documentation per §9.2.3, §9.2.4
- **Firm constraints:** Ground truth must exist BEFORE running extraction (§D3.1)
- **Flexible decisions:** Ground truth file format (JSON or YAML), location within demo app
- Manually trace all component-to-component connections in demo app to build ground truth
- Refactor demo app to implement EventPublisherDef for each event publisher
- Comparison script/test that diffs extraction output against ground truth on (source, target, type)
- Record timing in \`docs/architecture/performance/phase-12-baselines.md\`

### Embedded Reasoning
**Why:** Without ground truth validation, we cannot know if extraction is correct. This is the definitive proof that connection detection works end-to-end.
**What:** Ground truth definition, demo app refactor to Golden Path patterns, automated comparison test.
**How:** Manually audit demo app connections → write ground truth file → refactor publishEvent() → run extraction → diff against ground truth.

### Testing Strategy
- **Unit:** Ground truth file validates against schema.
- **Integration:** Full extraction pipeline run against demo app, output compared to ground truth. Zero diff = pass.
- **Edge Cases:** Transitive connections, events with no subscribers, multiple publishers for same event.

### Dependencies
Depends on #237 #238 #239 #240

### Verification
\`\`\`bash
pnpm nx test riviere-cli -- --testNamePattern "demo-app-validation"
# Or dedicated validation script
\`\`\`

## Summary by CodeRabbit

* **New Features**
  * Recognize event-publisher components; resolve publisher targets via publishedEventType metadata. Add parameter-type extraction with optional transforms. Support string-array literal values.

* **Documentation**
  * Added Phase 12 performance baselines and updated PRD notes for connection detection.

* **Tests**
  * Expanded suites for publish-connection resolution, parameter extraction, string[] literals, and extensive call-graph scenarios.

* **Chores**
  * CLI flag handling simplified to read command-line arguments directly.

### CODE REVIEW

File: packages/riviere-extract-ts/src/platform/domain/ast-literals/literal-detection.ts [hunk 1]
```diff
@@ -34,7 +34,8 @@ export function isLiteralValue(expression: Expression | undefined): boolean {
     kind === SyntaxKind.StringLiteral ||
     kind === SyntaxKind.NumericLiteral ||
     kind === SyntaxKind.TrueKeyword ||
-    kind === SyntaxKind.FalseKeyword
+    kind === SyntaxKind.FalseKeyword ||
+    kind === SyntaxKind.ArrayLiteralExpression
```
[the remark this answers is not in the document]
Comment (NTCoding) [reply packages/riviere-extract-ts/src/platform/domain/ast-literals/literal-detection.ts]: ✅ **Fixed**: isLiteralValue now validates array element types via isStringArrayLiteral helper — returns false for arrays with non-string elements (e.g., [42, 'foo']), matching extractLiteralValue behavior. Added test.

---
*Generated by Claude Code*

File: packages/riviere-extract-ts/src/domain/connection-detection/call-graph/trace-calls.ts [hunk 2]
```diff
@@ -10,10 +15,123 @@ import { componentIdentity } from './call-graph-types'
...
+    return directMatch
+  }
+
+  return resolveContainerMethod(
+    ctx.project,
+    resolvedTypeName ?? typeName,
+    calledMethodName,
+    ctx.componentIndex,
+  )
+}
```
[the remark this answers is not in the document]
Comment (NTCoding) [reply packages/riviere-extract-ts/src/domain/connection-detection/call-graph/trace-calls.ts]: ✅ **Fixed**: Changed resolveComponentTarget to return ComponentTargetOutcome with resolvedTypeName and uncertain. traceIntoNonComponent now consumes the pre-computed outcome instead of re-calling resolveTypeThroughInterface.

---
*Generated by Claude Code*

File: packages/riviere-extract-ts/src/domain/connection-detection/call-graph/trace-calls.ts [hunk 3]
```diff
@@ -10,10 +15,123 @@ import { componentIdentity } from './call-graph-types'
...
+    })
+  }
+}
+
+function traceBody(body: MethodDeclaration, ctx: TraceContext): void {
+  const callExpressions = body.getDescendantsOfKind(SyntaxKind.CallExpression)
+  for (const callExpr of callExpressions) {
+    traceCallExpression(callExpr, ctx)
+  }
+}
```
[the remark this answers is not in the document]
Comment (NTCoding) [reply packages/riviere-extract-ts/src/domain/connection-detection/call-graph/trace-calls.ts]: ❌ **Rejected**: traceBody and processFunction operate in fundamentally different processing contexts with different call expression handlers. traceBody uses traceCallExpression which operates within a TraceContext (transitive trace through non-components, with visited set and accumulated results). processFunction uses processCallExpression which handles top-level call processing with different uncertainty handling — it pushes uncertain links for unresolved types and delegates to traceNonComponent for the trace path. These are different code paths that share the same iteration pattern but have different semantics for how they handle type resolution failures, self-link detection, and uncertainty propagation. Widening the type would conflate two distinct processing phases, making the code harder to reason about.

---
*Generated by Claude Code*
